Implement multisort in DAGs table

mariana-marcal-santana opened the pull request:
Implement multisort for DAGs when displayed in a table; The URL displays extra sorting attributes when Shift+click is used on the column to sort by.

This implements the UI portion of #46383 and is dependent on the backend part implemented in #47440.
 
cc @pierrejeambrun @bbovenzi

mariana-marcal-santana committed:
Implement multisort in DAGs table
Frontend: Implement multisort for DAGs when displayed in a table;
The URL displays extra sorting attributes when Shift+click is used
on the column to sort by.

This implements the UI portion of #46383

pierrejeambrun reviewed (commented):
Thanks for your contribution.

Were you able to test this with a working backend ? (What I mean by that is you should be able to test that when you try sorting on multiple criteria the table should re-order itself appropriately.)

For this I believe you would need to rebase your branch on top of https://github.com/apache/airflow/pull/47440

pierrejeambrun reviewed (commented):
You need to modify components so that the request that is send to the backend passes the appropriate`order_by` list of query parameters. 

For now the url is correctly set and updated via the table component, but the requests send to the backend do not specify the multisort, this is why you do not have the expected response.

I think the backend implementation is functional and you should observe the appropriate behavior based on https://github.com/apache/airflow/pull/47440

mariana-marcal-santana committed:
Implement multisort in DAGs table
Frontend: Implement multisort for DAGs when displayed in a table;
The browser and backend URLs display extra sorting attributes when
Shift+click is used on the column to sort by.
This implements the UI portion of #46383

pierrejeambrun commented on airflow/ui/src/pages/DagsList/DagsList.tsx:
```diff
@@ -166,8 +166,7 @@ export const DagsList = () => {
 
-  const [sort] = sorting;
-  const orderBy = sort ? `${sort.desc ? "-" : ""}${sort.id}` : "-last_run_start_date";
+  const orderBy = new URLSearchParams(globalThis.location.search).getAll("sort");
```
You are missing the default sorting value.

By default if sort wasn't defined it would default to `-last_run_start_date`, that's lost with the new implementation

pierrejeambrun replied:
Also there are plenty of other places where we do sorting like that and that need to be updated to work with a list:
`AssetList`, `Variables`, `Connections` etc.... (you can search for `const [sort] = sorting` in the code base)

mariana-marcal-santana replied:
Thanks for the feedback, I'll start working on this asap!

pierrejeambrun reviewed (commented):
Thanks @mariana-marcal-santana,

Looking good, will do a final testing and review once the backend is merged. 

A few things need to be adjusted.

mariana-marcal-santana committed:
Implement multisort in DAGs table

Frontend: Implement multisort for DAGs when displayed in a table; Add default to the sorting parameters; Add support for multiple sort parameters in more lists

This implements the UI portion of apache#46383

bbovenzi commented on airflow/ui/src/pages/Asset/Asset.tsx:
```diff
@@ -37,8 +37,8 @@ export const Asset = () => {
-  const [sort] = sorting;
-  const orderBy = sort ? `${sort.desc ? "-" : ""}${sort.id}` : "-timestamp";
+  const orderBy = new URLSearchParams(globalThis.location.search).getAll("sort").length === 0 ?
+    ["-timestamp"] : new URLSearchParams(globalThis.location.search).getAll("sort");
```
```suggestion
  const sortParam = new URLSearchParams(globalThis.location.search).getAll("sort")
  const orderBy = sortParam.length === 0 ?
    ["-timestamp"] : sortParam;
```

bbovenzi replied:
I won't comment on every single occurrence  but let's try to keep our code DRY.

Since this is so common, maybe we can even pull it into its own util function or hook?

mariana-marcal-santana replied:
Hello @bbovenzi @pierrejeambrun !

I've created the function and used it in all places mentioned in https://github.com/apache/airflow/pull/47849#discussion_r2020629735.

I've also rebased the repo (yesterday) as requested.

I would appreciate any feedback on the implementation! Thanks!

bbovenzi reviewed (commented):
One quick comment and could you rebase when you have a chance?

After rebasing I'll do some more manual usability testing

mariana-marcal-santana committed:
Implement multisort in DAGs table
Frontend: Implement multisort for DAGs when displayed in a table;
Add function to get default sorting parameters

This implements the UI portion of apache#46383

mariana-marcal-santana committed:
Implement multisort in DAGs table
Frontend: Implement multisort for DAGs when displayed in a table;
Add support for multiple sort parameters in new lists

This implements the UI portion of apache#46383

bbovenzi commented on airflow-core/src/airflow/ui/src/utils/query.ts:
```diff
@@ -20,6 +20,11 @@ import { useDagServiceGetDagDetails } from "openapi/queries";
+  const sortParam = new URLSearchParams(globalThis.location.search).getAll("sort");
+  return (sortParam.length === 0 && pre_defined != undefined) ? [pre_defined] : sortParam;
+};
+
```
```suggestion
export const getOrderBy = (defaultOrder?: string): string[] => {
  const sortParam = new URLSearchParams(globalThis.location.search).getAll("sort");
  return (sortParam.length === 0 && defaultOrder !== undefined) ? [defaultOrder] : sortParam;
};

```

We generally don't want to use snake_case in our typescript code. The only ones we do have are generated from the python API.

mariana-marcal-santana replied:
Hello @bbovenzi !

In my last commit I fixed the casing, ran the afore mentioned commands and fixed the things marked in the lint command (that were from my commits).

Anything else I should change?

Also, I noticed that the code failed these tests: https://github.com/apache/airflow/actions/runs/14682302749/job/41221555783. How should I fix it? 
I noticed that it fails where I changed orderBy?: string to orderBy?: string[]. From what I saw, I think it's because it clashes with the definition on the file airflow-core/src/airflow/api_fastapi/core_api/openapi/v1-rest-api-generated.yaml (that I changed but I believe it got overwritten).

Could you confirm this please?

Thank you!

bbovenzi replied:
`pnpm format` should help solve most of those static checks

bbovenzi replied:
We may also need to make sure we generate the API spec docs before we run pnpm format

mariana-marcal-santana replied:
Hello @bbovenzi !

I ran pnpm format but the static checks still failed...

When you say API spec docs do you mean this file: airflow-core/src/airflow/api_fastapi/core_api/openapi/v1-rest-api-generated.yaml? And if so how can I generate it? I changed it locally, adapted the code that depended on it and everything seems to be working on my machine. What do I need to do to pass the static checks?

Thank you!

bbovenzi replied:
Oh I think this is an order of operations with pre-commit

`pre-commit run generate-openapi-spec` and then `pre-commit run ts-compile-format-lint-ui`

mariana-marcal-santana replied:
Hello @bbovenzi !

I had previously ran these operations and noticed that the first one changes the v1-rest-api-generated.yaml file like so:

![image](https://github.com/user-attachments/assets/b9414ebd-bee6-4545-a384-7e3b534c0d45)

This is what is causing the errors in the CI because the Order by argument now has to be an array and it keeps changing to string (which doesn't go along with the changes in the code for the multisort).
Given this, what I meant was if there was a way to define the file and the pre-commit to allow the Order By to be an array? From what I know, it seems like this could fix the CI errors.

Thank you!

bbovenzi reviewed (commented):
Looking good.

Do you mind running `pnpm lint && pnpm format` inside your local `airflow-core/src/airflow/ui` directory?

mariana-marcal-santana committed:
Implement multisort in DAGs table Frontend:
Implement multisort for DAGs when displayed in a table;
Fix casing and formatting

This implements the UI portion of apache#46383

pierrejeambrun commented on airflow-core/src/airflow/api_fastapi/core_api/openapi/v1-rest-api-generated.yaml:
```diff
@@ -2731,8 +2731,9 @@ paths:
-          default: dag_id
+          type: array
+          items:
+            type: string
```
You can't modify this manually, the file is auto generated based on the backend, and currently the backend specify a `string` not an array. You'll need to wait for the backend PR to be done and rebase on top of that, this way the `order_by` will have the appropriate `array` type.

Unfortunately the backend PR is not progressing. I trying pinging the contributor but we might want to take it at some point.

mariana-marcal-santana replied:
Hello @pierrejeambrun @bbovenzi !

Thanks for clarifying! Given this, should I fix the editions in this file now or only after the backend PR is merged to have a better sense of the necessary changes?

pierrejeambrun replied:
Once the backend PR is merged, rebasing on latest main will automatically give you that change. (The updated openapi spec).

We can't merge this PR until the backend is done, until then I think we can keep things like that (manually edited spec), this way from a `frontend` point of view the backend interface is correct and you can work without disrupting your workflow (eslint errors etc), but the CI static check won't be able to succeed. Once the backend is done we remove the manual change to the openapi spec and rebase on main to get the updated spec.

pierrejeambrun reviewed (commented):
@mariana-marcal-santana backend PR is done. You can rebase this one and we should be able to get it merged.

mariana-marcal-santana committed:
minor fixes

mariana-marcal-santana committed:
correction for assetlayout feature

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/Dag/Overview/Overview.tsx:
```diff
@@ -66,7 +66,7 @@ export const Overview = () => {
   const { data: gridRuns, isLoading: isLoadingRuns } = useGridRuns({ limit });
   const { data: assetEventsData, isLoading: isLoadingAssetEvents } = useAssetServiceGetAssetEvents({
-    limit,
+    limit: 6,
```
Why are we hardcoding "6" here?

pierrejeambrun reviewed (commented):
Were you able to test the feature?

Is that working for you. I'm not able to sort on multiple columns with that change. I believe we are missing a few pieces. (or maybe the rebase went wrong?)

Also CI need fixing.

pierrejeambrun reviewed (commented):
This PR went stale, closing
